app/models.py

```diff
@@ -2142,3 +2148,56 @@ def serialize(self):
             "created_at": created_at_in_bst.strftime("%Y-%m-%d %H:%M:%S"),
         }
         return contact_list
+
+
+class BroadcastMessage(db.Model):
+    __tablename__ = 'broadcast_message'
+    __table_args__ = (
+        db.ForeignKeyConstraint(
+            ['template_id', 'template_version'],
+            ['templates_history.id', 'templates_history.version'],
+        ),
+        {}
+    )
+
+    STATES = [
+        'draft',
+        'pending-approval',
+        'rejected',  # rejected before leaving notify
+        'broadcasting',  # approved, now live
+        'completed',  # reached end_date
+        'cancelled',  # the broadcast went out but then we sent a second cancel message to end it prematurely
+        'technical-failure',
+    ]
+
+    id = db.Column(UUID(as_uuid=True), primary_key=True)
+
+    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey('services.id'))
+    service = db.relationship('Service', backref='broadcast_messages')
+
+    template_id = db.Column(UUID(as_uuid=True), nullable=False)
+    template_version = db.Column(db.Integer, nullable=False)
+    template = db.relationship('TemplateHistory', backref='broadcast_messages')
+
+    _personalisation = db.Column(db.String, nullable=True)
+
+    state = db.Column(db.String)
+    CheckConstraint(f"state in ({','.join(STATES)})")
+
+    # these times are related to the actual broadcast, rather than auditing purposes
+    starts_at = db.Column(db.DateTime, nullable=True)
+    finishes_at = db.Column(db.DateTime, nullable=True)  # isn't updated if user cancels
+
+    # these times correspond to when
+    created_at = db.Column(db.DateTime, nullable=False)
+    approved_at = db.Column(db.DateTime, nullable=True)
+    cancelled_at = db.Column(db.DateTime, nullable=True)
```

Comment:
If cancelled is a terminating state do we need a separate `cancelled_at`?
There is no real harm in having the extra column, but I don't think a terminating state needs a separate timestamp

Reply:
hmmm, i didn't wanted to have to update `finishes_at` (which the users sets in a form) if the user clicks the cancel button though. i think it might still be useful for us to know "they originally planned for this to be active for three days, but stopped after one day". Or would you just rely on `updated_at` and checking the state? i guess that's not too bad actually

Reply:
Either way works. 
If you don't think `cancelled` is a terminating state we should keep the `cancelled_at` column. 
If you think it makes things a lot clearer that's reason enough to keep it.
Also means that some broadcast_messages will never set that column.

Reply:
i think that, based on our current understanding of the CAB model, `cancelled` should always be a terminating state. but i think i'll keep `cancelled_at`, as i think it'll help with clarity. i don't think we'll really know what the data should look like until we've sent a few messages (at least test messages) unfortunately